pkg/cloud/linode/provider.go

```diff
@@ -0,0 +1,195 @@
+package linode
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/linode/linodego"
+	"github.com/opencost/opencost/core/pkg/log"
+	"github.com/opencost/opencost/core/pkg/opencost"
+	"github.com/opencost/opencost/pkg/cloud/models"
+	"github.com/opencost/opencost/pkg/clustercache"
+	"github.com/opencost/opencost/pkg/env"
+	"golang.org/x/oauth2"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	InstanceAPIPricing = "Instance API Pricing"
+
+	kubeAPITimeout   = 30 * time.Second
+	linodeAPITimeout = 30 * time.Second
+)
+
+var initOnce = &sync.Once{}
+
+type LinodePricing map[string]*linodego.LinodeType
+
+type Linode struct {
+	KubeClient       kubernetes.Interface
+	Clientset        clustercache.ClusterCache
+	Config           models.ProviderConfig
+	ClusterRegion    string
+	ClusterAccountID string
+	ClusterProjectID string
+
+	linodeClient  atomic.Value
+	isHaCluster   atomic.Bool
+	regions       atomic.Value
+	linodePricing sync.Map
+}
+
+func (l *Linode) init() error {
+	cpricing, err := l.Config.GetCustomPricingData()
+	if err != nil {
+		log.Errorf("Could not get Linode custom pricing: %+v", err)
+
+		return err
+	}
+
+	clusterID, err := strconv.Atoi(strings.TrimLeft(l.ClusterProjectID, "lke"))
+	if err != nil {
+		log.Errorf("Could not parse LKE cluster ID (%s): %+v", l.ClusterProjectID, err)
+
+		return err
+	}
+
+	initOnce.Do(func() {
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+		go func() {
+			first := true
+
+			for {
+				linodeToken := os.Getenv("LINODE_TOKEN")
+				if cpricing.LinodeTokenSecret != "" && linodeToken == "" {
+					namespace := "default"
+					name := cpricing.LinodeTokenSecret
+					if secretName := strings.Split(cpricing.LinodeTokenSecret, "/"); len(secretName) > 1 {
+						namespace = secretName[0]
+						name = secretName[1]
+					}
+
+					ctx, cancel := context.WithTimeout(context.Background(), kubeAPITimeout)
+					if apiToken, err := l.KubeClient.CoreV1().Secrets(namespace).Get(ctx, name, v1.GetOptions{}); err != nil {
```

Comment:
@Sean-Holcomb So the main reason behind this solution is key rotation without downtime and extra operational actions. Linode doesn't fit into the static pricing model of OpenCost, so we fetch live pricing data from the provider. Because price should change, we fetch data in a loop, so I didn't see any reason to not refresh the token each time.

Reply:
My issue is that you are changing the signature of `provider.NewProvider()` to pass in the `kubeClientSet` which is a very large dependency to push through here. 

Just to be clear, you are suggesting that accessing the secret this way allows you to swap out the secret value without the pod coming down which mounting the secret which does not allow.